python/python/tests/test_util.py

```diff
@@ -84,3 +85,37 @@ def test_local_join_uri_windows():
         assert joined == str(pathlib.Path(base) / "table.lance")
         joined = join_uri(pathlib.Path(base), "table.lance")
         assert joined == pathlib.Path(base) / "table.lance"
+
+
+def test_value_to_sql_string(tmp_path):
+    # Make sure we can convert Python string literals to SQL strings, even if
+    # they contain characters meaningful in SQL, such as ' and \.
+    values = [
+        "anthony's",
+        "test\n newline",
+        "anthony's \n newline",
+    ]
+    expected_values = [
+        '"anthony\'s"',
+        '"test\n newline"',
+        '"anthony\'s \n newline"',
+    ]
```

Comment:
I'd still prefer to use `'` instead of `"` as SQL literals. Generally `'` is standard SQL for strings. I think it's only an accident that our SQL parser accepts double quotes.

I think you are correct the newlines don't need special escaping. So we can remove those.

Instead, we should add cases that mix single and double quotes. That seems to accurately point out the weakness of your implementation. Instead, I think what you want to do is

```python
value = value.replace("'", "\'")
return f"'{value}"'
```

```suggestion
    values = [
        "anthony's",
        "anthony's \"favorite color\" wasn't red"
    ]
    expected_values = [
        "'anthony\'s'",
        "'anthony\'s \"favorite color\" wasn\'t red",
    ]
```

Reply:
I have tried what you suggested, the problem is anytime I try to wrap a value in single-quotes, there is an sql parser error from lance-datafusion->sqlparser->tokenizer.

The only case where the parsing is successful is when expected values are as follows:
```python
    values = ["anthony's", "anthony's \"favorite color\" wasn't red"]
    expected_values = [
        '"anthony\'s"',
        '"anthony\'s "favorite color" wasn\'t red"',
    ]  # fmt: skip
```

i.e. wrapping in double quotes. If you feel this should work with single quotes, then I think there's some problem with upstream lance-datafusion implementation. I tried looking at lance-datafusion, but in order to check that I need to go through datafusion repository and sqlparser repo, for which I need a bit of time. Another genuine possibility is, I am entirely clueless.

Reply:
The second case (`"'anthony\'s \"favorite color\" wasn\'t red'"`) doesn't pass no matter what. I have tried manually passing this in the `updates` table in the following line: [lance/python/python/lance/dataset.py#L1090](https://github.com/lancedb/lance/blob/67c984f10a63ce3889d92d4d968ccc5180455fd5/python/python/lance/dataset.py#L1090), but none of the combinations work.